pkg/activator/podip_watcher.go

```diff
@@ -0,0 +1,265 @@
+					prober.ExpectsBody(queue.Name))
+				healthyDests[dest] = ok
+
+				if err != nil {
+					rw.logger.Errorw(fmt.Sprintf("Probing of dest %q failed", dest), zap.Error(err))
+				} else if !ok {
+					rw.logger.Infof("Probing of dest %q failed.", dest)
+				}
+
+				probeWg.Done()
+			}(dest)
+		}
+	}
+
+	probeWg.Wait()
+
+	rw.healthyDestsMux.Lock()
+	defer rw.healthyDestsMux.Unlock()
+	rw.healthyDests = healthyDests
+}
+
+func (rw *revisionWatcher) runOnce(tickCh <-chan time.Time) bool {
+	select {
+	case dests := <-rw.destsChan:
+		if dests == nil {
+			// shutdown
+			return true
+		}
+		rw.dests = dests
+		rw.checkDests()
+	case <-tickCh:
+		rw.checkDests()
+	}
+	return false
+}
+
+func (rw *revisionWatcher) run() {
+	ticker := time.NewTicker(200 * time.Millisecond)
+	defer ticker.Stop()
+
+	for {
+		if rw.runOnce(ticker.C) {
+			return
+		}
+	}
+}
+
+// PodIPWatcher listens to revision endpoints and keeps track of healthy
+// l4 dests which can be used to reach a revisoin
+type PodIPWatcher struct {
+	revisionWatchers    map[string]*revisionWatcher
+	revisionDestsCh     map[string]chan []string
+	revisionWatchersMux sync.RWMutex
+
+	transport http.RoundTripper
+	logger    *zap.SugaredLogger
+}
+
+// NewPodIPWatcher returnes a PodIPWatcher
+func NewPodIPWatcher(transport http.RoundTripper,
+	endpointsInformer corev1informers.EndpointsInformer, logger *zap.SugaredLogger) *PodIPWatcher {
+	piw := &PodIPWatcher{
+		revisionWatchers: make(map[string]*revisionWatcher),
+		revisionDestsCh:  make(map[string]chan []string),
+		transport:        transport,
+		logger:           logger,
+	}
+
+	endpointsInformer.Informer().AddEventHandler(cache.FilteringResourceEventHandler{
+		FilterFunc: reconciler.ChainFilterFuncs(
+			reconciler.LabelExistsFilterFunc(serving.RevisionUID),
+			// We are only interested in the private services, since that is
+			// what is populated by the actual revision backends.
+			reconciler.LabelFilterFunc(networking.ServiceTypeKey, string(networking.ServiceTypePrivate), true),
+		),
+		Handler: cache.ResourceEventHandlerFuncs{
+			AddFunc:    piw.endpointsUpdated,
+			UpdateFunc: controller.PassNew(piw.endpointsUpdated),
+			DeleteFunc: piw.endpointsDeleted,
+		},
+	})
+
+	return piw
+}
+
+// GetHealthyPodDests returns a list of l4 dests which are "healthy"
+// (have returned a success on the most recent probe)
+func (pw *PodIPWatcher) GetHealthyPodDests(rev RevisionID) []string {
+	pw.revisionWatchersMux.RLock()
+	defer pw.revisionWatchersMux.RUnlock()
+
+	// We dont need to create if one does not exist so not using getOrCreateRevisionWatcher
+	// This also allow us to only use a read lock
+	rw, ok := pw.revisionWatchers[rev.String()]
+	if !ok {
+		return nil
+	}
+	return rw.getHealthyDests()
+}
+
+func (pw *PodIPWatcher) getOrCreateDestsCh(rev RevisionID) chan []string {
+	revStr := rev.String()
+	destsCh, ok := pw.revisionDestsCh[revStr]
+	if !ok {
+		destsCh = make(chan []string)
```

Comment:
do we want the channel to be buffered?

Reply:
I dont think so - this should be out of any synchronous request path (just async informer updates) so IMO blocking is a lot nicer